File: 199.cpp
/*
Given a binary tree, imagine yourself standing on the right side of it, return the values of the nodes you can see ordered from top to bottom.

For example:
Given the following binary tree,
   1            <---
 /   \
2     3         <---
 \     \
  5     4       <---
You should return [1, 3, 4].
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* root, int lv, vector<int> &res){
        if(!root)   return;
        if(lv>=res.size()) res.push_back(root->val);
        dfs(root->right,lv+1,res);
        dfs(root->left,lv+1,res);
    }
    
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        dfs(root, 0, res);
        return res;
    }
    
    vector<int> rightSideView(TreeNode* root) {
        queue<TreeNode*> q;
        vector<int> res;
        
        if (!root) return res;
        q.push(root);
        while (!q.empty()) {
            int n = q.size();
            res.push_back(q.front()->val);
            while (n--) {
                auto cur = q.front();
                q.pop();
                
                if (cur->right) q.push(cur->right);
                if (cur->left) q.push(cur->left);
            }
        }
        return res;
    }
};

File: 199.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    // O(1) solution
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> results = new ArrayList<>();
        dfs(root, 0, results);
        return results;
    }
    
    private void dfs(TreeNode root, int level, List<Integer> results) {
        if (root == null) return;
        if (level == results.size())
            results.add(root.val);
        dfs(root.right, level+1, results);
        dfs(root.left, level+1, results);
    }
}
